Maid.lua

```diff
@@ -1,52 +1,76 @@
-local MakeMaid do
-	local index = {
-		GiveTask = function(self, Task)
-			local n = #self.Tasks + 1
-			self.Tasks[n] = Task
-			return n
-		end;
-		
-		DoCleaning = function(self)
-			local Tasks = self.Tasks
-			for Name, Task in next, Tasks do
-				if type(Task) == "function" then
-					Task()
-				else
-					Task:Disconnect()
-				end
-				Tasks[Name] = nil
-			end
-		end;
-	};
-	index.Disconnect = index.DoCleaning -- Allow maids to be stacked.
+-- Maid[key] = (function)            Adds a function to call at cleanup
+-- Maid[key] = (Instance)            Adds an Object to be Destroyed at cleanup
+-- Maid[key] = (RBXScriptConnection) Adds a connection to be Disconnected at cleanup
+-- Maid[key] = (Maid)                Maids can act as an event connection, allowing a Maid to have other maids to clean up.
+-- Maid[key] = nil                   Removes a named task. This cleans up the previous Maid[key]
+local Maid = {}
 
-	local mt = {
-		__index = function(self, k)
-			if index[k] then
-				return index[k]
-			else
-				return self.Tasks[k]
-			end
-		end;
-		
-		__newindex = function(self, k, v)
-			local Tasks = self.Tasks
-			if v == nil then
-				if type(Tasks[k]) ~= "function" and Tasks[k] then
-					Tasks[k]:Disconnect() -- disconnect if the task is an event
-				end
-			elseif Tasks[k] then
-				self[k] = nil -- clear previous task
-			end
-			Tasks[k] = v
-		end;
-	}
+--- Generates a new Maid object
+-- @return Maid
+function Maid.new()
+	return setmetatable({{}}, Maid)
+end
+
+--- Gives the Maid a Task to perform at cleanup, incremented by 1
+-- @param Variant Task An object to be Destroyed, a Connection to be Disconnected, or function/callable table to be called
+-- @returns the index the Task was placed at
+function Maid:GiveTask(Task)
+	local n = #self[1] + 1
+	self[1][n] = Task
+	return n
+end
+
+--- Makes the Maid clean up when the instance is destroyed
+-- @param Instance Instance The Instance the Maid will wait for to be Destroyed
+function Maid:LinkToInstance(Instance)
+	self:GiveTask(Instance.AncestryChanged:Connect(function(Object, Parent)
+		if Instance == Object and Parent == nil then
```

Comment:
You don't want to check if `Instance == Object` because that means you won't get notified when `Instance.Parent.Parent = nil` Also you shouldn't use `Parent` you should just call `Instance:IsDescendantOf(game)` which covers the same edge case of `Instance.Parent.Parent = nil`. 

Finally, this doesn't cover if the object is just simply unparented than reparented. This is a long term issue. I wouldn't worry too much about this. You may have an issue with objects parented to characters (tools have this issue). You can fix this by adding a `wait()` but that's disgusting for a number of reasons. 

I'm not sure what the performance cost is, but a work around may exist with `CollectionService`. I'll need to look into this myself.